tracing: optimize hot path

sitano opened the pull request:
Results
===

Saves about 7% on the hot path best case for pure in-memory prepared statements executions: from around 12% to 5%.

Tested on C2 16 vCPU single shard Scylla pure in-memory reads with 100% cache hit rate. On production, most of the operations will be much heavier and the relative performance degradation will be even smaller as far as we target 10,000 QPS / shard max and users don't run shards at 100% utilization normally.

Performance
===

Commits | Title | Perf Regression | Perf Saving
-- | -- | -- | --
a8c90a584 | master - without tracing | 0.00% |  
a8c90a584 | master - with slow tracing | 33.52% |  
a8c90a584 | master - with fast tracing | 11.58% | 21.95%
00becd70f | tracing: remove tracing::begin(lazy_eval) | 9.29% | 2.28%
a92883106 | tracing: add begin(type), deprecate begin(sstring) | 8.78% | 0.51%
e9b5b0135 | tracing: ctx params inline vectors | 8.22% | 0.56%
de100df96 | tracing: avoid copying a query string in a prepared statements tracing hot path | 7.61% | 0.61%
1708cc4a9 | tracing: make client_state.user to be lw_shared_ptr | 5.64% | 1.97%

Another potential optimization points
===

- find out why flattening `struct trace_state::params_value` by removing the `struct trace_state::params_ptr` does not show performance benefit hence removes an unnecessary lazy allocation
- move `tracing::add_prepared_query_options(trace_state, options);` out to the end of the request processing behind `tracing::trace(q_state->query_state.get_trace_state(), "Done processing - preparing a result");`. 

Other improvements
===

- rewrite budget handling to be added and subtracted from the same code paths. remove size() dependency on ready(). maybe remove ready().
- can remove `backend_session_state_base` and just move the whole content of the impl to the base.


Scripts
===

- [reads.sh.txt](https://github.com/sitano/scylla/files/6223565/reads.sh.txt)

Commit:
tracing: remove tracing::begin(lazy_eval)

To optimize code hot paths for tracing we are trying to remove dynamic
allocations and memory copying as much as possible. One of the biggest
contributor to the performance overhead in tracing is string formatting.
There was only 1 usage of this form of begin and the only reason it
existed was to delay argument formatting until the implementation
checked that the trace_state is initialized. For the fast slow query
mode we are trying to omit doing dyn allocs and copies of the values.

This patch gracefully sacrificies the flexibility of this call in favor
of better performance.

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

Commit:
tracing: add trace_string_enum type for begin(type)

This patch prepares move from the tracing::begin(string) to
use of enum that encodes all of the types of the tracing contexts
with an integer that is much cheaper to store.

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

Commit:
tracing: add begin(type), deprecate begin(sstring)

To optimize code hot paths for tracing we are trying to remove dynamic
allocations and memory copying as much as possible. One of the biggest
contributor to the performance overhead in tracing is string formatting.

With this patch we will move from tracing::begin that accepts strings to
tracing::begin(enum) that encodes the type of the tracing context with
an integer that is much cheaper to store.

This patch introduces a tracing::begin(enum) and updates all call sites.
It changes the way tracing context stores the title of the tracing
context to the usage of enum.

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

Commit:
tracing: avoid copying a query string in a prepared statements tracing hot path

To optimize code hot paths for tracing we are trying to remove dynamic
allocations and memory copying as much as possible. One of the biggest
contributor to the performance overhead in tracing is string formatting.

The hottest code path is expected to be the execution of prepared
statements. A prepared statement usually has a single query to trace.
Usually prep statements are all compiled and cached. Thus we can omit
copying and allocating memory of keeping the track of the related query.

This patch changes the way trace state params tracks queries for cached
prepared statement by reusing shared_ptr. This allows to save 1 dyn
allocation and copy on the hot path.

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

Commit:
tracing: ctx params inline vectors

To optimize code hot paths for tracing we are trying to remove dynamic
allocations and memory copying as much as possible. One of the biggest
contributor to the performance overhead in tracing is string formatting.

Not all tracing sessions are gonna to be written but all of them
fulfilled with the context parameters. There is a set of parameters that
is always filled. In most of the code hot paths there will be a query
and parameters to save. Most of the queries we expect to be non batch,
have 1 query and at most 2 parameters. That's why we can save a few
allocations here by changing the usage of std::vector to small_vector's.

This patch introduces use of small_vectors to minimize num of allocs in
code hot paths. small_vectors use local storage for storing objects
until their capacity is enough.

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

vladzcloudius commented on transport/server.cc:
```diff
@@ -989,7 +989,7 @@ process_query_internal(service::client_state& client_state, distributed<cql3::qu
     auto skip_metadata = options.skip_metadata();
 
-    if (init_trace) {
+    if (init_trace && trace_state) {
```
I don't like this patch: we already have `if (trace_state_ptr)` branch inside of every public `tracing::xx()` function.

I don't see how this patch can deliver any significant improvement when Slow Query Fast is enabled. 
I do see how it can make things worse however.

Note that this patch optimized the "tracing disabled" case on account of "tracing enabled" case.

So, it's counterproductive given the goal of this series.

sitano replied:
@vladzcloudius you are right that we have `if (trace_state_ptr)` inside. I just thought that we don't want to do multiple branch checking when we can omit doing that. you are also correct that it does not correspond to the purpose of this PR - its just a hot path optimization. I also don't understand why we have those `trace_state` setter private? we could have that public.

do you want me to remove this patch?

sitano replied:
removed

vladzcloudius commented on tracing/trace_string.cc:
```diff
@@ -0,0 +1,54 @@
+
+namespace tracing {
+
+const char *trace_string_data[] = {
```
use `std::small_vector<sstring>` instead of C-arrays.

sitano replied:
ok, but this would require adding 2 includes `<vector>` and `"sstring"` for nothing.

sitano replied:
reworked

sitano replied:
now with operators

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,45 @@
+        return trace_string_data[static_cast<int>(val)];
+    }
+
+    inline bool is_set(trace_string_enum val) {
```
Why is this needed?

sitano replied:
originally I wanted to have a boolean operator overload on the enum class. but turned out to be it's impossible to have functions inside of the enum class. there are workarounds but I did not want it. the default initialization of the enum class is zero-initialization however zero value does not mean it's a default value for an enum class. so I wanted to have a method that explicitly says what is the default value for the type.

it is used in this context
```
if (request.empty() && ::tracing::trace_string::is_set(record.request_type)) {
        request = std::string_view(::tracing::trace_string::to_string(record.request_type));
}
```

otherwise I will have to rewrite it into `&& static_cast<int>(record.request_type)`.

vladzcloudius replied:
I think boolean operator is a better choice.

sitano replied:
now with operators

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,45 @@
+extern const char* trace_string_data[];
+
+namespace trace_string {
+    inline const char *to_string(trace_string_enum val) {
```
Implement an `operator*(enum class trace_string_enum)` instead.

The you can get the string value by calling `*enum_val` which is nicer.

sitano replied:
ok, but I don't see any advantage from having this logic encapsulated into an operator*. using to_string function is more explicit variant for a reader.

vladzcloudius replied:
The advantage is a more laconic (hence more readable) call site.

sitano replied:
now with operator*

sitano commented:
I have done manual testing in Debug and Release mode with tracing trace enabled for slow, fast/slow, probablistic, tracing on, tracing off modes. It works. Fixed 1 bug with missing username and table name.

@vladzcloudius I have added this since 31 march 2021:

```
void trace_state::set_session_records_params() noexcept {
    _records->session_rec.client = _client;
    _records->session_rec.request = _request;
    _records->session_rec.request_type = _request_type;
    _records->session_rec.started_at = _started_at;
    _records->session_rec.request_size = _request_size;
    _records->session_rec.response_size = _response_size;

    if (_params_ptr) {
        _records->session_rec.username = std::move(_params_ptr->username);

        for (const auto &table: _params_ptr->tables) {
            _records->session_rec.tables.emplace(table);
        }
    }
}
```

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -288,6 +289,10 @@ cql3::query_options trace_keyspace_helper::make_session_mutation_data(const one_
     parameters_values_vector.reserve(record.parameters.size());
     std::for_each(record.parameters.begin(), record.parameters.end(), [&parameters_values_vector] (auto& val_pair) { parameters_values_vector.emplace_back(val_pair.first, val_pair.second); });
     auto my_map_type = map_type_impl::get_instance(utf8_type, utf8_type, true);
+    std::string_view request = record.request;
```
Why is this needed? The whole point is to get rid of records.request and the corresponding copy, no?

sitano replied:
removed with begin(string)

vladzcloudius commented on tracing/tracing.hh:
```diff
@@ -218,6 +220,7 @@ struct session_record {
     std::set<sstring> tables;
     sstring username;
     sstring request;
+    trace_string_enum request_type = {};
```
Don't you want to initialize it to `none`?

sitano replied:
moved to default type constructor

vladzcloudius commented on tracing/tracing.hh:
```diff
@@ -218,6 +220,7 @@ struct session_record {
     std::set<sstring> tables;
     sstring username;
     sstring request;
```
Drop `request`, please.

sitano replied:
dropped, removed with begin(string)

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -281,6 +281,7 @@ private:
      * @param request description of a request being traces
      * @param client address of a client the traced request came from
      */
```
I think we should just kill it.
If somebody wants to be verbose - he/she should do it using tracing::trace(...) which allows a full blown formatting.

sitano replied:
killed

vladzcloudius commented on test/boost/tracing.cc:
```diff
@@ -29,12 +30,10 @@
 
 future<> do_with_tracing_env(std::function<future<>(cql_test_env&)> func, cql_test_config cfg_in = {}) {
     return do_with_cql_env([func](auto &env) {
-        // supervisor::notify("creating tracing");
```
How is this related to this patch?
Don't mix unrelated cleanups with the functional patch.
Send this as a separate patch if needed.

sitano replied:
extracted

vladzcloudius commented on test/boost/tracing.cc:
```diff
@@ -29,12 +30,10 @@
         tracing::register_tracing_keyspace_backend(tracing_backend_registry);
         tracing::tracing::create_tracing(tracing_backend_registry, "trace_keyspace_helper").get();
 
-        // supervisor::notify("starting tracing");
```
same here

sitano replied:
extracted

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -61,6 +61,7 @@ struct trace_state::params_values {
     std::optional<db::consistency_level> cl;
     std::optional<db::consistency_level> serial_cl;
     std::optional<int32_t> page_size;
+    // TODO: move to the shared_ptr to ensure we keep queries
```
This comment is confusing. 
Please, remove it.

sitano replied:
removed

vladzcloudius reviewed (changes requested):
I don't understand why this patch is needed and why do you need anything except for catching the shared_ptr as described in that comment.

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -213,9 +217,11 @@ void trace_state::build_parameters_map() {
-        const prepared_checked_weak_ptr& prepared_ptr = prepared_statements[i];
-        build_parameters_map_for_one_prepared(prepared_ptr, vals.query_option_names[i], vals.query_option_values[i], param_key("param", i, psize <= 1));
-        params_map.emplace(param_key("query", query_key_ix, single_key), prepared_ptr ? prepared_ptr->statement->raw_cql_statement : "[lost]");
+        build_parameters_map_for_one_prepared(prepared_statements[i], vals.query_option_names[i], vals.query_option_values[i], param_key("param", i, psize <= 1));
```
I'm not sure I understand why you needed to change the loop so much while all what's changed is the place you get the query string from.

sitano replied:
because there were special case when there is only 1 val and when there are X vals, and I needed X+1 val.

sitano replied:
rework  to disconnect semantical parts

sitano replied:
reworked

vladzcloudius reviewed (commented):
Why only 2 allocations? I see at least 5.

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -94,16 +95,23 @@ public:
     trace_state_props_set _state_props;
     state _state = state::inactive;
-    std::chrono::system_clock::rep _started_at;
-    gms::inet_address _client;
```
Why did you need to move this and others up and down?
Please, separate functional changes from the cleanups. 
It makes it very hard to understand what is really going on here.

sitano replied:
because those fields logically are touched by different pieces of code compared to the 

```
    trace_state_props_set _state_props;
    state _state = state::inactive;
    ```

sitano replied:
moved to flat _records

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -94,16 +95,23 @@ public:
 private:
     lw_shared_ptr<one_session_records> _records;
+
+    trace_type _command = trace_type::NONE;
```
Please, arrange these in a nice struct so that it would be very clear that these are going to eb copied into the one_session_records.

sitano replied:
Why it's not clear? 

I can flatten params then I move everything there.

sitano replied:
moved to flat _records

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -485,7 +485,7 @@ future<> trace_keyspace_helper::flush_one_session_mutations(lw_shared_ptr<one_se
                         return _sessions_time_idx.insert(qp, _dummy_query_state, make_session_time_idx_mutation_data, std::ref(*records));
                     }).then([this, &qp, records] {
-                        if (!records->do_log_slow_query) {
+                        if (records->session_rec.elapsed < _local_tracing.slow_query_threshold()) {
```
This changes the semantics of the slow_query_threshold.
Please, send changes related to this as a separate patch outside this series.
I doubt this change delivers any measurable performance gain.

sitano replied:
> This changes the semantics of the slow_query_threshold.

how?

> Please, send changes related to this as a separate patch outside this series.

Then the patch will not work because I have deleted this parameter from the records.

sitano replied:
this is probably a bug by me because elapsed will not be set for non primary records that still have to be saved. also as Vlad correctly mentioned on other nodes thresholds may be not configured.

sitano replied:
thanks its not a bug because this part of the code writes only the final session to the node_slow table. however, I am returning it back as it is does not relate to this patch.

sitano replied:
previosly, records->do_log_slow_query depended only on local non serialized tracing.slow_query_threshold() so that code does not change semantics

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -54,9 +54,14 @@ namespace tracing {
     template<typename T, size_t N> using small_vector = ::utils::small_vector<T, N>;
 
+    sstring username;
+    utils::small_vector<sstring, 1> tables;
```
Why these lines are needed?

sitano replied:
because we postpone creation of the one_session_record.

sitano replied:
moved to flat _records

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -94,16 +95,23 @@ public:
 
 private:
     lw_shared_ptr<one_session_records> _records;
```
Why, instead of this whole patch not to store `one_session_record` as a field in the trace_state and create a copy of it only when needed - when you pass it to a different layer (e.g. to the I/O backend).

This way you'll achieve exactly same (actually probably better) result without needed to do that many changes.

sitano replied:
because `one_session_record` has different lifetime characteristics and also is part of the tracing record, not part of the tracing context. context exists temporary, `one_session_record` exists beyond that is may outlive at least during postponed writes. See https://github.com/sitano/scylla/pull/2/commits/62b1475f5c75f5f706ffa25a1a19b2bfad718202#diff-bddf710626ec338c0b779c234871530ed27a052367dc4018f44b55ea2e534c22R359.

sitano replied:
moved to flat _records

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -158,17 +171,22 @@ public:
+        if (_records) {
+            return _records->session_id;
+        } else {
+            return default_session_id;
```
This can only work for a Slow Logging Fast tracing mode where you never ask for a tracing on the remote node.
In all other modes you need to pass the real UUID to the remote side.

sitano replied:
no, there is 1 use case that is too general - the tracing error assertion in the state_trace destructor - maybe we could delete it?

> https://github.com/sitano/scylla/pull/2/commits/62b1475f5c75f5f706ffa25a1a19b2bfad718202#diff-bddf710626ec338c0b779c234871530ed27a052367dc4018f44b55ea2e534c22R270-R273

sitano replied:
moved to flat _records

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -531,7 +531,6 @@ private:
     friend void add_table_name(const trace_state_ptr& p, const sstring& ks_name, const sstring& cf_name);
     friend void add_prepared_query_options(const trace_state_ptr& state, const cql3::query_options& prepared_options_ptr);
     friend void stop_foreground(const trace_state_ptr& state) noexcept;
-    friend void stop_foreground_prepared(const trace_state_ptr& state, const cql3::query_options* prepared_options_ptr) noexcept;
```
Where did this go?

sitano replied:
I don't know. It's a dead statement.

sitano replied:
extracted no a diff patch

vladzcloudius reviewed (changes requested):
Again, you are trying to optimize a non-Tracing path.
Why do you care?

sitano commented:
@vladzcloudius next round please

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,47 @@
+
+extern std::vector<seastar::sstring> trace_string_data;
+
+inline const seastar::sstring &operator*(trace_string_enum val) {
```
Some cosmetics:
* "&" should be attached to the type (Avi's code style) like this: `const sstring&`
* No need to use the `seastar::sstring` - include `seastarx.hh` and it is going to introduce a `seastar` namespace into your scope.

sitano replied:
done

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,47 @@
+extern std::vector<seastar::sstring> trace_string_data;
+
+inline const seastar::sstring &operator*(trace_string_enum val) {
+    return trace_string_data[static_cast<int>(val)];
```
No range check. Better use `at()`

sitano replied:
Ok

sitano replied:
sadly saying it's impossible to write a static_assert against 

- enum size
- vector size - non const

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,47 @@
+    return trace_string_data[static_cast<int>(val)];
+}
+
+inline bool operator!(trace_string_enum val) {
```
Better implement `operator bool`.
See more on "why" here: https://en.cppreference.com/w/cpp/language/operators

sitano replied:
There is no trivial way to implement this operator Vlad for enum class. I mean there is no way at all for enum classes. There are only workarounds that are not beautiful. That only add to code complexity. But ok. I will replace my simple ! operator with class+enum+bool.

sitano replied:
introduced 
```
class trace_string_enum
{
public:
    enum value_type : uint8_t
    {
        none = 0,
        prepare_stmt,
        execute_stmt,
        execute_prepared,
        execute_batch,
        alternator
    };

    trace_string_enum() = default;
    constexpr trace_string_enum(value_type type) : value(type) { }

    operator value_type() const { return value; }

    explicit operator const sstring&() const {
        return trace_string_data.at(value);
    }

    [[nodiscard]]
    bool empty() const { return !value; }

private:
    value_type value;
}
```

that is of course much more clean than a single enum class with 1 func.

sitano replied:
My best effort to keep enum and strings as close as possible:
```
namespace tracing {

namespace string {

    enum value_type : uint8_t
    {
        none = 0,
        prepare_stmt,
        execute_stmt,
        execute_prepared,
        execute_batch,
        alternator
    };

    // the number of elements must match value_type enum
    inline std::vector<sstring> trace_string_data = {
        "None",
        "Preparing CQL3 query",
        "Execute CQL3 query",
        "Execute CQL3 prepared query",
        "Execute batch of CQL3 queries",
        "Alternator"
    };

}

class trace_string_enum
{
public:
    using value_type = tracing::string::value_type;

    trace_string_enum() = default;
    constexpr trace_string_enum(value_type type) : value(type) { }

    explicit operator bool() const { return empty(); }

    sstring operator*() const {
        return string::trace_string_data.at(value);
    }

    [[nodiscard]]
    bool empty() const { return !value; }

private:
    value_type value;
};
```

> trace_string_enum() = default;

default empty constructor is required to be able to write

```
trace_string_enum();
trace_string_enum a = {};
trace_string_enum b = trace_string_enum();
or anything alike
```

sadly saying it's impossible to write a static_assert against

-    enum size
-    vector size - non const

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -288,6 +289,10 @@ cql3::query_options trace_keyspace_helper::make_session_mutation_data(const one_
     std::for_each(record.parameters.begin(), record.parameters.end(), [&parameters_values_vector] (auto& val_pair) { parameters_values_vector.emplace_back(val_pair.first, val_pair.second); });
     auto my_map_type = map_type_impl::get_instance(utf8_type, utf8_type, true);
+    std::string_view request = record.request;
+    if (request.empty() && !!record.request_type) {
```
Here you go.
If you had an `operator::bool` you wouldn't need this `!!` nonsense.

sitano replied:
There is no trivial way to implement this operator Vlad for `enum class`. I mean there is no way at all for enum classes. There are only workarounds that are not beautiful. That only add to code complexity. But ok. I will replace my simple ! operator with class+enum+bool.

sitano replied:
done by operator

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -288,6 +289,10 @@ cql3::query_options trace_keyspace_helper::make_session_mutation_data(const one_
     parameters_values_vector.reserve(record.parameters.size());
     std::for_each(record.parameters.begin(), record.parameters.end(), [&parameters_values_vector] (auto& val_pair) { parameters_values_vector.emplace_back(val_pair.first, val_pair.second); });
     auto my_map_type = map_type_impl::get_instance(utf8_type, utf8_type, true);
+    std::string_view request = record.request;
```
I would squash the patches that introduce the `begin(type)` and remove all other `begin(...)` into a single patch.

sitano replied:
Ok

sitano replied:
squashed

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -195,6 +203,13 @@ void trace_state::build_parameters_map() {
+        size_t psize = statements.size();
+        for (size_t i = 0; i < psize; ++i) {
+            const prepared_checked_weak_ptr& prepared_ptr = statements[i];
+            params_map.emplace(param_key("query", i, psize <= 1), prepared_ptr ? prepared_ptr->statement->raw_cql_statement : "[lost]");
```
Not good.
We want to always have a query string.
Please, always catch the `prepared->statement` when a `tracing::add_prepared_statement(...)` is called.

sitano replied:
I have no idea how do you want me this to be done. 

Y:ou did not like it when it was:

```
- keys_size = query.size() + statements.size(), keys_index = 0
- add all queries(keys_size, keys_index)
- add all statements(keys_size, keys_index)
```

You don't like:

```
- if queries: add all queries()
- else: add all statements()
```

I don't see other options. Please, write the preudocode of what you want here. 

I am deleting my helper to let this code be left as bloated as it is (sstring param_key(const char *key, size_t index, bool single) {).

The new variant is:

```
    if (!prepared_statements.empty()) {
        // Parameter's key in the map will be "param[X]" for a single query CQL command and "param[Y][X] for a multiple
        // queries CQL command, where X is an index of the parameter in a corresponding query and Y is an index of the
        // corresponding query in the BATCH.
        if (prepared_statements.size() == 1) {
            build_parameters_map_for_one_prepared(prepared_statements[0].prepared, vals.query_option_names[0], vals.query_option_values[0], "param");
        } else {
            // BATCH
            for (size_t i = 0; i < prepared_statements.size(); ++i) {
                build_parameters_map_for_one_prepared(prepared_statements[i].prepared, vals.query_option_names[i], vals.query_option_values[i], format("param[{:d}]", i));
            }
        }

        const size_t qsize = queries.size();
        for (size_t i = 0; i < psize; ++i) {
            params_map.emplace(param_key("query", qsize+i, qsize+psize), prepared_statements[i].cql->raw_cql_statement);
        }
    }
```

sitano replied:
extracting into separate patch

sitano replied:
replaced with 
```
    auto& statements = vals.prepared_statements;
    const size_t psize = statements.size();
    if (psize) {
        // Parameter's key in the map will be "param[X]" for a single query CQL command and "param[Y][X] for a multiple
        // queries CQL command, where X is an index of the parameter in a corresponding query and Y is an index of the
        // corresponding query in the BATCH.
        for (size_t i = 0; i < psize; ++i) {
            build_parameters_map_for_one_prepared(statements[i].prepared, vals.query_option_names[i], vals.query_option_values[i], param_key("param", i, psize));
            params_map.emplace(param_key("query", qsize+i, qsize+psize), statements[i].cql->raw_cql_statement);
        }
    }
    ```

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -161,6 +161,14 @@ void trace_state::add_prepared_query_options(const cql3::query_options& prepared
     }
 }
 
+sstring param_key(const char *key, size_t index, bool single) {
```
How this is related to what this patch claims to do?

This patch is supposed to catch the `prepared->statement` shared_ptr and then use `statement->cql_raw_query` instead of `request.query`. 

Please, introduce a `param_key` beautifier as a separate patch.

sitano replied:
ok, I will dequeue this completelly. It is too much work to keep such changes separate Vlad.

sitano replied:
removed

sitano replied:
recovered into separate patch

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -61,7 +61,11 @@ struct trace_state::params_values {
     std::optional<db::consistency_level> serial_cl;
     std::optional<int32_t> page_size;
+    // provides an access to the cql3 statements of present prepared statements
+    std::vector<shared_ptr<cql3::cql_statement>> cql_statements;
```
It's very bad that you create a state of the tree when there is a point where Tracing is broken (after the previous patch).

Please, re-arrange this and the previous patches as follows:
1) One that introduces `params_key`: definition and usage.
2) One that changes copying the prepared statement query string to catching the shared_ptr.

sitano replied:
ok, I will just merge htem.

sitano replied:
merged

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -145,6 +149,7 @@ void trace_state::set_user_timestamp(api::timestamp_type val) {
 
 void trace_state::add_prepared_statement(prepared_checked_weak_ptr& prepared) {
     _params_ptr->prepared_statements.emplace_back(prepared->checked_weak_from_this());
+    _params_ptr->cql_statements.emplace_back(prepared->statement);
```
What if the second `emplace` throws?

sitano replied:
Nothing bad will happen besides the loss of the statement. There are so many things that can throw in the middle and we will not get data. For would you suggest? Replacing it with a struct? Wrapping it try/catch? I don't see what will improve this.

sitano replied:
replaced with struct that however does not change anything if emplace will throw

sitano replied:
replaced with a single struct

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -61,7 +61,11 @@ struct trace_state::params_values {
     std::optional<db::consistency_level> cl;
     std::optional<db::consistency_level> serial_cl;
     std::optional<int32_t> page_size;
+    // provides an access to the cql3 statements of present prepared statements
```
Better have a single vector of structs than multiple vectors when both vectors are always appended together.

sitano replied:
intro-ed statement_param

vladzcloudius commented on transport/server.cc:
```diff
@@ -478,11 +478,10 @@ future<foreign_ptr<std::unique_ptr<cql_server::response>>>
         }
     }
 
-    tracing::set_request_size(trace_state, fbuf.bytes_left());
```
Again, this is not going to optimize a Tracing hot path but rather a not-Tracing hot path.

Please, exclude these from this series and send them as a follow up.

sitano replied:
It's too hard Vlad to make all those changes in separate patches. I will just dequeue it. On the other hand. It is a separate commit, right?

sitano replied:
dropped

vladzcloudius commented on transport/server.cc:
```diff
@@ -559,7 +556,11 @@ future<foreign_ptr<std::unique_ptr<cql_server::response>>>
             }
 
-            tracing::set_response_size(trace_state, response->size());
+            if (trace_state && (trace_state->full_tracing() || trace_state->should_log_slow_query(trace_state->elapsed()))) {
```
This is bogus: what if `trace_state->should_log_slow_query(trace_state->elapsed()))` returns you `false` here but it would return true during the following call of `tracing::stop_foreground(trace_state)`?

It will result in `trace_state` missing data.

sitano replied:
I don't see @vladzcloudius how is that bogus.

1. This patch gave me about 1-1.5% of win in our test by saving dynamic formatting and allocating/coping username. I would not recheck that tho.
2. This `if` check is the latest coroutine in the sequence of request processing. At that point deferred stop_foreground call will be the next instruction right to it. So there will be NO data loss until an exception and it is possible to neglect the possibility of the data lose because its limit is 0.
3.  I thought of calculating _do_log_slow_query at this point and passing it to the stop_foreground call but that's too many changes for too little win. I rather rollback this patch than invest my time into it.

However, as I understand you dislike this optimization, so I will dequeue this patch without further ado.

sitano replied:
dropped

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -561,8 +566,17 @@ inline void trace_state::trace_internal(sstring message) {
+            copy->events_recs = std::move(_records.events_recs);
+            _records.events_recs = {};
+            copy->parent_id = _records.parent_id;
+            copy->my_span_id = _records.my_span_id;
```
Implement a copy-constructor for a `one_session_records`, please.
But even that is not going to be good enough because one_session_records contains not only a static data but also contains a state (e.g. `_is_pending_for_write`) and is also used to access and modify the Tracing state via `budget_ptr` and the later is a rather non-trivial since I dynamically change the `budget_ptr` depending in which state the current one_session_records object is right now. Namely the specific `budget_ptr` represents a part of the `one_session_records` object state.

I have another idea: I think the main reason for a performance penalty here is a dereference operation, so instead of flattening `one_session_records` you can try to cache references to corresponding fields:

```
trace_state {
....
std::vector<sstring>& _queries;
```
and then 
```
trace_state::trace_state(...) 
: _queries(_params_ptr->queries)
...
```

and then you'll have 
```
void trace_state::add_query(sstring_view val) {
    _queries.emplace_back(std::move(val));
}
```

I suspect this would deliver you the majority of the benefit.

sitano replied:
> Implement a copy-constructor for a one_session_records, please.

I can just merge my 2 patches so I will not have to keep this intermediate code and implement copy con.

> But even that is not going to be good enough because one_session_records contains not only a static data but also contains a state (e.g. _is_pending_for_write) and is also used to access and modify the Tracing state via budget_ptr and the later is a rather non-trivial since I dynamically change the budget_ptr depending in which state the current one_session_records object is right now. Namely the specific budget_ptr represents a part of the one_session_records object state.

Yes, you are correct. It's unfortunate it tracks both parameters and the state so I can't just copy it and forget it. I have to ensure that the budget will be properly tracked and I do that. The main issue with the budget tracking is that its returning is done outside of the main budget withdrawing code paths. And moreover, that the amount that is withdrawn depends on the Ready() state of the object. I thought to fix that, but now I think its too much toil for me to do it. 

> I have another idea: I think the main reason for a performance penalty here is a dereference operation, so instead of flattening one_session_records you can try to cache references to corresponding fields.

I don't know. The dereference is absent when I flattened `unique_ptr<one_session_records>` but the perf lowered. This is counter-intuitive. We expect it to be the opposite. 

What you shown in the code however:

```
trace_state {
....
std::vector<sstring>& _queries;
```

 is not used in our prep statement code path (_queries). What did you mean?

sitano replied:
testing simplified version perf

sitano commented:
@vladzcloudius next round. please, address the questions. benchmarks on the way.

Commit:
tracing: refactoring of params serialization

just introduces a local helper param_key to
simplify evaluating of the parameters names

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

sitano commented:
alternative https://github.com/sitano/scylla/tree/tracing_optimize_hot_path6 without removing unique_ptr allocation of one_record on hot path

Commit:
tracing: make client_state.user to be lw_shared_ptr

to save a memory allocation and string formatting during tracing for the
username.

To optimize code hot paths for tracing we are trying to remove dynamic
allocations and memory copying as much as possible. One of the biggest
contributor to the performance overhead in tracing is string formatting.

User authentication object exists in the lifetime context of the
connection. A user does an authentication only once. Thus we can omit
copying this object in the hot path of query execution by reusing
a shared ptr to it.

This patch changes an option to lw_shared_ptr for auth user context to
keep only the ptr instead of the value inside of the trace state params.

Signed-off-by: Ivan Prisyazhnyy <ivan@scylladb.com>

sitano commented:
new alternative https://github.com/sitano/scylla/tree/tracing_optimize_hot_path7
- no patch to omit one_session_record allocation and session_id and span_id generation
- username to lw_shared_ptr in client_state

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,74 @@
+
+namespace tracing {
+
+namespace string {
```
tracing::string::XX - this is cool!

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,74 @@
+    trace_string_enum() = default;
+    constexpr trace_string_enum(value_type type) : value(type) { }
+
+    explicit operator bool() const { return empty(); }
```
Do you really want `bool` to be explicit?

sitano replied:
usual practice not to make bool conversion implicit because it may then implicitly convert to integer

sitano replied:
removed

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,74 @@
+
+namespace string {
+
+    enum value_type : uint8_t
```
`enum class`, no?

sitano replied:
if I will do that enum class, then it will be not `tracing::string::begin` but `tracing::string::value_type::begin`

sitano replied:
my best enum I have every created
```
+enum class string_type : uint8_t
+{
+    none = 0,
+    prepare_stmt,
+    execute_stmt,
+    execute_prepared,
+    execute_batch,
+    alternator,
+    last
+};
+
+inline std::array<const char *, 1+static_cast<size_t>(string_type::last)> trace_string_data = {
+    "None",
+    "Preparing CQL3 query",
+    "Execute CQL3 query",
+    "Execute CQL3 prepared query",
+    "Execute batch of CQL3 queries",
+    "Alternator",
+    "Last"
+};
+
+// none == 0 is the default value.
+static_assert(static_cast<size_t>(string_type::none) == 0);
+
+class trace_string_enum
+{
+public:
+    trace_string_enum() = default;
+    constexpr trace_string_enum(string_type type) : value(type) { }
+
+    sstring operator*() const {
+        return trace_string_data[static_cast<size_t>(value)];
+    }
+
+    [[nodiscard]]
+    bool empty() const { return value == string_type::none; }
+
+private:
+    string_type value;
+};
```

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    [[nodiscard]]
+    bool empty() const { return !value; }
```
here better use `value == value_type::none`. 
It's, first, more readable and more robust - what if `value_type::none` stops being 0 in the future?

Another way to ensure that this doesn't break is to add a static assert that would check that `none == 0`

sitano replied:
good point

sitano replied:
done

vladzcloudius reviewed (changes requested):
Please, add a note in the patch description that you are removing the prepared statement ID in the process.

vladzcloudius reviewed (changes requested):
Same here. This patch introduces a new thing - put some more text about it.

vladzcloudius reviewed (approved):
Very nice!

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -163,6 +163,14 @@ void trace_state::add_prepared_query_options(const cql3::query_options& prepared
     }
 }
 
+static sstring param_key(const char *key, size_t index, size_t size) {
```
Why not to make it a private method of a `trace_state`?

Also, add an JavaDoc description for this new function/method, please.

sitano replied:
static makes it already private enough to the translation unit. no need in making it part of the trace_state declaration. no need in declaring it in the header.

vladzcloudius replied:
Oh, I see... Makes sense.

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -214,7 +224,8 @@ void trace_state::build_parameters_map() {
         for (size_t i = 0; i < psize; ++i) {
-            build_parameters_map_for_one_prepared(statements[i], vals.query_option_names[i], vals.query_option_values[i], param_key("param", i, psize));
+            build_parameters_map_for_one_prepared(statements[i].prepared, vals.query_option_names[i], vals.query_option_values[i], param_key("param", i, psize));
+            params_map.emplace(param_key("query", qsize+i, qsize+psize), statements[i].cql->raw_cql_statement);
```
Nitpicks: unlike in Python Avi's codding style requires spaces around `+`: like this:
```
params_map.emplace(param_key("query", qsize + i, qsize + psize), statements[i].cql->raw_cql_statement);
```

And the description of the patch is not the best - this patch is more about saving a string copy in the tracing hot path than just saving a function call. The reader wouldn't understand that one leads to another - you should to be explicit.

Except for that - very nice!

sitano replied:
done

vladzcloudius commented on transport/server.cc:
```diff
@@ -477,11 +477,10 @@ future<foreign_ptr<std::unique_ptr<cql_server::response>>>
             cqlop == cql_binary_opcode::BATCH) {
             trace_props.set_if<tracing::trace_state_props::write_on_close>(tracing_request == tracing_request_type::write_on_close);
             trace_state = tracing::tracing::get_local_tracing_instance().create_session(tracing::trace_type::QUERY, trace_props);
+            tracing::set_request_size(trace_state, fbuf.bytes_left());
```
How is this patch related to this series?
It optimized a non-tracing path.

Please, send it as a follow up patch.

sitano replied:
it is a separate patch

vladzcloudius replied:
I mean a separate patch NOT in this PR.
Send it as a follow up (after this PR).

sitano replied:
extracted out

vladzcloudius commented on cql3/query_processor.hh:
```diff
@@ -177,7 +177,7 @@ public:
     }
 
-    statements::prepared_statement::checked_weak_ptr get_prepared(const std::optional<auth::authenticated_user>& user, const prepared_cache_key_type& key) {
+    statements::prepared_statement::checked_weak_ptr get_prepared(const lw_shared_ptr<auth::authenticated_user>& user, const prepared_cache_key_type& key) {
```
About the patch description: here, again, it's not about formatting it's about string copy in a hot path that you postpone now.

So, the current description of the patch doesn't help understanding the right motivation.

sitano replied:
I think now it's better

vladzcloudius commented on test/boost/cql_auth_query_test.cc:
```diff
@@ -69,13 +69,13 @@ static void create_user_if_not_exists(cql_test_env& env, std::string_view user_n
 
-    const auto reset = defer([&cs, old_user = std::move(old_user)] {
-        cs.set_login(std::move(*old_user));
+    const auto reset = defer([&cs, old_user] {
```
This is not going to work as expected.
Original lines are restoring the user name as it was before the `f()` call which probably modifies the value of `cs.user()`. 

So, by catching the shared pointer to it you are not going to achieve the original goal.
You need to catch the *value* of the original user, namely `[..., old_user = *cs.user()]`

sitano replied:
I agree that in that specific case its better to catch by val but it works here because when the test calls `cs.set_login(auth::authenticated_user(sstring(user_name)));` it changes the place where the `client_state._user` points to. So that `old_user` pointer is untouched.

vladzcloudius replied:
This is the internal implementation detail. In the future someone may decide to change it so that when `set_login()` is called the value is assigned and not the whole shared_pointer is re-allocated and the test will break.

sitano replied:
yes, this is fixed

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -302,6 +303,7 @@ cql3::query_options trace_keyspace_helper::make_session_mutation_data(const one_
         "username",
         "ttl"
     };
+    std::strstream username; username << *record.user ;
```
I don't mind but you may get a comment about not breaking a line after `;`.

sitano replied:
breaked

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -313,7 +315,7 @@ cql3::query_options trace_keyspace_helper::make_session_mutation_data(const one_
         cql3::raw_value::make_value(int32_type->decompose((int32_t)(record.request_size))),
         cql3::raw_value::make_value(int32_type->decompose((int32_t)(record.response_size))),
-        cql3::raw_value::make_value(utf8_type->decompose(record.username)),
+        cql3::raw_value::make_value(utf8_type->decompose(sstring(username.str()))),
```
do you really need an explicit cast to `sstring` here?

sitano replied:
no, I misinterpreted decompose semantics

sitano replied:
removed

sitano replied:
replaced with helper
```
+static sstring user_to_string(const lw_shared_ptr<auth::authenticated_user>& user) {
+    std::strstream username;
+    if (user) {
+        username << *user;
+    } else {
+        username << auth::anonymous_user();
+    }
+    return sstring(username.str());
+}
```

vladzcloudius commented on tracing/tracing.hh:
```diff
@@ -210,14 +211,16 @@ struct event_record {
         , event_time_point(event_time_point_) {}
 };
 
+inline lw_shared_ptr<auth::authenticated_user> _default_unauthenticated_user = make_lw_shared(auth::authenticated_user());
```
This is not going to work.
lw_shared_ptr can only live on the same shard it's been allocated on.
So, in order to achieve what you want to achieve here you need to create a non-static private field in one of the per-shard tracing objects, like `tracing::tracing` and then use it  in a default constructor.

sitano replied:
lol, this is my bad. data race on the flat surface.

sitano replied:
removed completely

sitano commented:
@vladzcloudius all comments addressed. next round pls.

vladzcloudius commented on alternator/server.cc:
```diff
@@ -355,7 +355,7 @@ static tracing::trace_state_ptr maybe_trace_query(service::client_state& client_
         std::string buf;
```
Paragraph 2 and 3 in the description are telling the same (very similar) thing but in different words. I think you should drop one of them.

sitano replied:
fixed

vladzcloudius commented on tracing/trace_state.cc:
```diff
@@ -54,6 +54,13 @@ namespace tracing {
 
 logging::logger trace_state_logger("trace_state");
```
Change the subject of the patch, please. - It doesn't express what this patch really is about.
I think something like "tracing: avoid copying a query string in a prepared statements tracing hot path" would be clearer.

sitano replied:
changed

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -390,9 +390,9 @@ private:
     void add_prepared_statement(prepared_checked_weak_ptr& prepared);
 
-    void set_username(const std::optional<auth::authenticated_user>& user) {
+    void set_user(const lw_shared_ptr<auth::authenticated_user>& user) {
```
better change to passing the parameter by value and use std::move during the assignment.
This is a more modern C++ way.

sitano replied:
changed to move

vladzcloudius commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -281,6 +282,16 @@ void trace_keyspace_helper::write_records_bulk(records_bulk& bulk) {
     });
 }
 
+static sstring user_to_string(const lw_shared_ptr<auth::authenticated_user>& user) {
```
I wonder why you need all this?
The original code has an assignment from auth::authenticated_user to sstring. 

And most importantly you are breaking the abstraction: trace_keyspace_helper should have no idea about the trace_state related logic.

here the change should be as simple as:

```
-        cql3::raw_value::make_value(utf8_type->decompose(record.username)),
+        cql3::raw_value::make_value(utf8_type->decompose(sstring(*record.user))),
```

But even this is not good enough because "unathenticated request is not the same as a request that is authenticated with an `anonymous_user`.

So, you may want to implement a `trace_state::get_username_string()` method that would preserve the current functionality and have the logic similar to below inside.

sitano replied:
can't  move it to trace_state. `user_to_string` moved to `session_record::get_username_string()`

vladzcloudius commented on tracing/trace_state.hh:
```diff
@@ -649,9 +649,9 @@ inline void add_prepared_statement(const trace_state_ptr& p, prepared_checked_we
 }
 
-inline void set_username(const trace_state_ptr& p, const std::optional<auth::authenticated_user>& user) {
+inline void set_user(const trace_state_ptr& p, const lw_shared_ptr<auth::authenticated_user>& user) {
```
Same here - pass shared_pointers by value, please, and use std::move()

sitano replied:
passed

vladzcloudius commented on transport/server.cc:
```diff
@@ -682,7 +682,7 @@ client_data cql_server::connection::make_client_data() const {
     cd.driver_version = _client_state.get_driver_version();
-    if (const auto user_ptr = _client_state.user(); user_ptr) {
+    if (const auto &user_ptr = _client_state.user(); user_ptr) {
         cd.username = user_ptr->name;
```
Hold on! Here is your username string! It's freaking `public`!

vladzcloudius commented on test/boost/cql_auth_query_test.cc:
```diff
@@ -69,13 +69,13 @@ static void create_user_if_not_exists(cql_test_env& env, std::string_view user_n
 static void with_user(cql_test_env& env, std::string_view user_name, noncopyable_function<void ()> f) {
     auto& cs = env.local_client_state();
-    std::optional<auth::authenticated_user> old_user = cs.user();
+    auto old_user = *(cs.user());
```
ditch brackets

sitano replied:
yep

vladzcloudius commented on tracing/trace_string.hh:
```diff
@@ -0,0 +1,77 @@
+
+    // trace_string_enum.empty() expects none == 0.
+    static_assert(none == 0);
+
```
remove null-string

sitano replied:
removed

sitano commented:
@vladzcloudius all comments addressed. next round pls.
